### PR TITLE
[WFCORE-3521]: Display dynamic capabilities with a pattern.

### PR DESCRIPTION
**Pull request by ehsavoie**
Jira: https://issues.jboss.org/browse/WFCORE-3521

This is helpful for provisioning

### CODE REVIEW

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this pease

File: controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java [hunk 1]
```diff
@@ -215,8 +229,8 @@ String getDependentName(RuntimeCapability cap, OperationContext context) {
             if (cap.isDynamicallyNamed()) {
                 return cap.fromBaseCapability(context.getCurrentAddress()).getName();
             } else {
-                return cap.getName();
-            }
+            return cap.getName();
+        }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java]: This whitespace change looks incorrect.

File: controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java [hunk 2]
```diff
@@ -167,4 +172,17 @@ private ModelNode addDeprecatedInfo(final ModelNode model) {
...
+        while (iter.hasNext()) {
+            PathElement elt = iter.next();
+            if (elt.isWildcard()) {
+                aliasAddress = aliasAddress.append(PathElement.pathElement(elt.getKey(), "$" + elt.getKey()));
+            } else {
+                aliasAddress = aliasAddress.append(elt);
+            }
+        }
+        return aliasAddress;
+    }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java]: Please document this. It's quite clever. :) But the cleverness makes it confusing.

File: controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java [hunk 3]
```diff
@@ -76,6 +76,21 @@ default boolean isDynamicDependent() {
...
+     * dynamicElements the dynmaic elements of the requirement capability name.
+     * @return requirement name of the capability this reference depends on
+     */
+    default String getRequirementName(String... dynamicElements) {
+        if (dynamicElements != null && dynamicElements.length > 0) {
+            String[] dynamicParts = new String[dynamicElements.length];
+            for (int i = 0; i < dynamicElements.length; i++) {
+                dynamicParts[i] = "$" + dynamicElements[i];
+            }
+            return RuntimeCapability.buildDynamicCapabilityName(getBaseRequirementName(), dynamicParts);
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java]: It's better to not repeat the static part. We've gotten complaints before about the size of our metadata.

**Comment by bstansberry** [comment]
https://github.com/wildfly/wildfly/compare/master...bstansberry:cap-req-doc is a draft doc with possible names for this metadata.

Also, we should avoid outputting this in the standard case, again to cut the size of our descriptions.

Perhaps the capability ref recorder should return List<String> instead of String and then the consumer deals with the $ part. That way if you pass in 'foo' and it replies with [`foo`] it can efficiently detect that and skip the output.

File: controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java [hunk 4]
```diff
@@ -76,6 +76,21 @@ default boolean isDynamicDependent() {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * dynamicElements the dynmaic elements of the requirement capability name.
+     * @return requirement name of the capability this reference depends on
+     */
+    default String getRequirementName(String... dynamicElements) {
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java]: i wasn't sure at first, but this is not required

File: controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java (hunk 3 above)
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java]: ok

File: controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java [hunk 5]
```diff
@@ -103,6 +105,9 @@ public ModelNode getModelDescription(Locale locale) {
             for (Capability capability: capabilities) {
                 ModelNode cap = result.get(ModelDescriptionConstants.CAPABILITIES).add();
                 cap.get(NAME).set(capability.getName());
+                if(capability.isDynamicallyNamed()) {
+                    cap.get(DYNAMIC_NAME).set(capability.getDynamicName(createAliasPathAddress(pa)));
+                }
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java]: ok

**Comment by bstansberry** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

File: controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java [hunk 6]
```diff
@@ -76,6 +77,23 @@ default boolean isDynamicDependent() {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Returns the elements to be added to the baseRequirementName to build the capability name pattern.
+     * The pattern is of the form `$segment[.$segment]` where each segment represents either the name of one of the
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java]: Not: The form is now `segment[.segment]`. The RRD output is now a LIST of STRING with the $ char no longer relevant.

File: controller/src/main/java/org/jboss/as/controller/AttributeDefinition.java [hunk 7]
```diff
@@ -982,8 +982,15 @@ public ModelNode getNoTextDescription(boolean forOperation) {
                 result.get(ModelDescriptionConstants.REQUIRES).add(required);
             }
         }
+
         if (getReferenceRecorder() != null) {
+            String[] capabilityPatternElements = getReferenceRecorder().getRequirementPatternSegments(name);
             result.get(ModelDescriptionConstants.CAPABILITY_REFERENCE).set(getReferenceRecorder().getBaseRequirementName());
+            if(capabilityPatternElements.length > 0) {
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/AttributeDefinition.java]:
```
if(capabilityPatternElements.length > 0 && (capabilityPatternElements.length > 1 || !name.equals(capabilityPatternElements[0])) { 
```

IOW don't output this for the default case.

If not outputting this makes the feature-spec generator too difficult let me know.

File: controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java [hunk 8]
```diff
@@ -99,11 +104,18 @@ public ModelNode getModelDescription(Locale locale) {
...
         Set<? extends Capability> capabilities = registration.getCapabilities();
-        if (capabilities!=null&&!capabilities.isEmpty()){
+        if (capabilities !=null && !capabilities.isEmpty()){
             for (Capability capability: capabilities) {
                 ModelNode cap = result.get(ModelDescriptionConstants.CAPABILITIES).add();
                 cap.get(NAME).set(capability.getName());
                 cap.get(DYNAMIC).set(capability.isDynamicallyNamed());
+                if(capability.isDynamicallyNamed()) {
+                    String dynamicPart = capability.getDynamicName(createAliasPathAddress(pa)).substring(capability.getName().length());
+                    StringTokenizer tokenizer = new StringTokenizer(dynamicPart, ".$", false);
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java]: If you use String.split you can get an array and again test for the case of a 1 element array whose element is the same as pa.getLastElement().getKey(). Don't output more description data for the default case.

**Comment by ehsavoie** [comment]
retest this please

File: controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java (hunk 8 above)
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/descriptions/DefaultResourceDescriptionProvider.java]: Looks like it is not breaking my usage

**Pull request merged**